File: python/energy_nft_minter.py
#!/usr/bin/env python3
"""
에너지 인증서 NFT 자동 민팅 스크립트
RE100, CBAM, 탄소거래용 NFT 생성 및 관리
"""

import json
import os
import requests
import time
from datetime import datetime
from typing import Dict, Any, Optional
from web3 import Web3
from eth_account import Account
import logging

# 로깅 설정
logging.basicConfig(level=logging.INFO, format='%(asctime)s - %(levelname)s - %(message)s')
logger = logging.getLogger(__name__)

class EnergyNFTMinter:
    def __init__(self, config_path: str = "config.json"):
        """
        에너지 NFT 민터 초기화
        Args:
            config_path: 설정 파일 경로
        """
        self.config = self.load_config(config_path)
        self.w3 = self.setup_web3()
        self.contract = self.setup_contract()
        self.account = Account.from_key(self.config['private_key'])
        
    def load_config(self, config_path: str) -> Dict[str, Any]:
        """설정 파일 로드"""
        try:
            with open(config_path, 'r', encoding='utf-8') as f:
                config = json.load(f)
            logger.info("✅ 설정 파일 로드 완료")
            return config
        except FileNotFoundError:
            logger.error("❌ 설정 파일을 찾을 수 없습니다. config.json을 생성해주세요.")
            raise
            
    def setup_web3(self) -> Web3:
        """Web3 연결 설정"""
        w3 = Web3(Web3.HTTPProvider(self.config['rpc_url']))
        if not w3.is_connected():
            raise ConnectionError("❌ 블록체인 네트워크에 연결할 수 없습니다.")
        logger.info(f"✅ {self.config['network']} 네트워크 연결 완료")
        return w3
        
    def setup_contract(self) -> Any:
        """스마트컨트랙트 설정"""
        try:
            with open(self.config['contract_abi_path'], 'r') as f:
                abi = json.load(f)
            
            contract = self.w3.eth.contract(
                address=self.config['contract_address'],
                abi=abi
            )
            logger.info("✅ 스마트컨트랙트 연결 완료")
            return contract
        except Exception as e:
            logger.error(f"❌ 컨트랙트 설정 실패: {e}")
            raise
            
    def generate_metadata(self, energy_data: Dict[str, Any]) -> Dict[str, Any]:
        """
        에너지 거래 데이터를 기반으로 NFT 메타데이터 생성
        Args:
            energy_data: 에너지 거래 정보
        Returns:
            NFT 메타데이터 딕셔너리
        """
        # 탄소 회피량 계산 (태양광: 0.466 tCO2/MWh, 풍력: 0.011 tCO2/MWh)
        carbon_factors = {
            'solar': 0.466,
            'wind': 0.011,
            'hydro': 0.024,
            'biomass': 0.18
        }
        
        carbon_offset = energy_data['energy_amount'] * carbon_factors.get(
            energy_data['energy_type'], 0.4
        )
        
        metadata = {
            "name": f"RE100 전력 NFT #{energy_data.get('token_id', 'TBD')}",
            "description": f"{energy_data['energy_amount']}MWh {energy_data['energy_type']} 전력 거래 인증서",
            "image": self.get_energy_type_image(energy_data['energy_type']),
            "external_url": f"https://energy-nft.com/token/{energy_data.get('token_id', '')}",
            "attributes": [
                {
                    "trait_type": "전력량 (MWh)",
                    "value": energy_data['energy_amount'],
                    "display_type": "number"
                },
                {
                    "trait_type": "에너지 타입",
                    "value": energy_data['energy_type']
                },
                {
                    "trait_type": "거래일시",
                    "value": energy_data.get('timestamp', datetime.now().isoformat())
                },
                {
                    "trait_type": "공급자",
                    "value": energy_data['supplier']
                },
                {
                    "trait_type": "수요자",
                    "value": energy_data['buyer']
                },
                {
                    "trait_type": "발전소 위치",
                    "value": energy_data['location']
                },
                {
                    "trait_type": "탄소 회피량 (tCO₂)",
                    "value": round(carbon_offset, 3),
                    "display_type": "number"
                },
                {
                    "trait_type": "RE100 적합성",
                    "value": "인증됨"
                },
                {
                    "trait_type": "CBAM 대응",
                    "value": "적합"
                }
            ],
            "properties": {
                "category": "Energy Certificate",
                "subcategory": energy_data['energy_type'].title(),
                "is_renewable": True,
                "certification_standard": "RE100",
                "carbon_offset_tco2": carbon_offset,
                "energy_amount_mwh": energy_data['energy_amount']
            }
        }
        
        return metadata
        
    def get_energy_type_image(self, energy_type: str) -> str:
        """에너지 타입별 이미지 URL 반환"""
        image_mapping = {
            'solar': 'ipfs://QmSolarPanelIcon/solar.png',
            'wind': 'ipfs://QmWindTurbineIcon/wind.png',
            'hydro': 'ipfs://QmHydroIcon/hydro.png',
            'biomass': 'ipfs://QmBiomassIcon/biomass.png'
        }
        return image_mapping.get(energy_type, 'ipfs://QmDefaultEnergyIcon/default.png')
        
    def upload_to_ipfs(self, metadata: Dict[str, Any]) -> str:
        """
        메타데이터를 IPFS에 업로드 (Pinata 서비스 사용)
        Args:
            metadata: 업로드할 메타데이터
        Returns:
            IPFS 해시 (ipfs:// 형식)
        """
        try:
            # 임시 파일로 메타데이터 저장
            filename = f"metadata_{int(time.time())}.json"
            with open(filename, 'w', encoding='utf-8') as f:
                json.dump(metadata, f, ensure_ascii=False, indent=2)
            
            # Pinata API를 통해 IPFS 업로드
            url = "https://api.pinata.cloud/pinning/pinFileToIPFS"
            headers = {
                'pinata_api_key': self.config['pinata_api_key'],
                'pinata_secret_api_key': self.config['pinata_secret_key']
            }
            
            with open(filename, 'rb') as f:
                files = {'file': f}
                response = requests.post(url, files=files, headers=headers)
            
            # 임시 파일 삭제
            os.remove(filename)
            
            if response.status_code == 200:
                ipfs_hash = response.json()['IpfsHash']
                ipfs_url = f"ipfs://{ipfs_hash}"
                logger.info(f"✅ IPFS 업로드 완료: {ipfs_url}")
                return ipfs_url
            else:
                raise Exception(f"IPFS 업로드 실패: {response.text}")
                
        except Exception as e:
            logger.error(f"❌ IPFS 업로드 오류: {e}")
            raise
            
    def mint_nft(self, energy_data: Dict[str, Any]) -> str:
        """
        에너지 NFT 민팅
        Args:
            energy_data: 에너지 거래 데이터
        Returns:
            트랜잭션 해시
        """
        try:
            logger.info("🚀 NFT 민팅 시작...")
            
            # 1. 메타데이터 생성
            metadata = self.generate_metadata(energy_data)
            logger.info("📄 메타데이터 생성 완료")
            
            # 2. IPFS 업로드
            token_uri = self.upload_to_ipfs(metadata)
            
            # 3. 탄소 회피량 계산
            carbon_factors = {'solar': 0.466, 'wind': 0.011, 'hydro': 0.024, 'biomass': 0.18}
            carbon_offset = int(energy_data['energy_amount'] * carbon_factors.get(
                energy_data['energy_type'], 0.4
            ) * 1000)  # 소수점 3자리까지 정수로 변환
            
            # 4. 스마트컨트랙트 함수 호출 준비
            recipient = energy_data['buyer']
            energy_amount = int(energy_data['energy_amount'] * 1000)  # MWh를 kWh로 변환
            supplier = energy_data['supplier']
            buyer = energy_data['buyer']
            energy_type = energy_data['energy_type']
            location = energy_data['location']
            
            # 5. 트랜잭션 빌드
            nonce = self.w3.eth.get_transaction_count(self.account.address)
            
            transaction = self.contract.functions.mintEnergyNFT(
                recipient,
                token_uri,
                energy_amount,
                supplier,
                buyer,
                energy_type,
                carbon_offset,
                location
            ).build_transaction({
                'from': self.account.address,
                'nonce': nonce,
                'gas': 500000,
                'gasPrice': self.w3.to_wei('10', 'gwei')
            })
            
            # 6. 트랜잭션 서명 및 전송
            signed_txn = self.w3.eth.account.sign_transaction(transaction, self.config['private_key'])
            tx_hash = self.w3.eth.send_raw_transaction(signed_txn.rawTransaction)
            
            logger.info(f"✅ 민팅 트랜잭션 전송 완료: {self.w3.to_hex(tx_hash)}")
            
            # 7. 트랜잭션 확인 대기
            receipt = self.w3.eth.wait_for_transaction_receipt(tx_hash)
            
            if receipt.status == 1:
                logger.info(f"🎉 NFT 민팅 성공! Gas 사용량: {receipt.gasUsed}")
                return self.w3.to_hex(tx_hash)
            else:
                raise Exception("트랜잭션 실패")
                
        except Exception as e:
            logger.error(f"❌ NFT 민팅 실패: {e}")
            raise
            
    def batch_mint(self, energy_data_list: list) -> list:
        """
        배치 민팅 (여러 NFT 한 번에 민팅)
        Args:
            energy_data_list: 에너지 데이터 리스트
        Returns:
            트랜잭션 해시 리스트
        """
        results = []
        for i, energy_data in enumerate(energy_data_list):
            try:
                logger.info(f"📦 배치 민팅 진행 중... ({i+1}/{len(energy_data_list)})")
                tx_hash = self.mint_nft(energy_data)
                results.append({'success': True, 'tx_hash': tx_hash, 'data': energy_data})
                
                # API 레이트 리미트 방지를 위한 대기
                time.sleep(2)
                
            except Exception as e:
                logger.error(f"❌ 배치 민팅 {i+1}번 실패: {e}")
                results.append({'success': False, 'error': str(e), 'data': energy_data})
                
        return results


def main():
    """메인 실행 함수"""
    try:
        # NFT 민터 초기화
        minter = EnergyNFTMinter("config.json")
        
        # 예시 에너지 거래 데이터
        sample_energy_data = {
            "energy_amount": 1.5,  # 1.5 MWh
            "energy_type": "solar",
            "supplier": "0x742d35Cc6634C0532925a3b8D7e99B8fF4b4c4C7",
            "buyer": "0x8ba1f109551bD432803012645Hac136c903C2B3F",
            "location": "경기도 화성시 태양광발전소",
            "timestamp": datetime.now().isoformat()
        }
        
        # NFT 민팅 실행
        tx_hash = minter.mint_nft(sample_energy_data)
        print(f"🎉 NFT 민팅 완료! 트랜잭션 해시: {tx_hash}")
        
    except Exception as e:
        logger.error(f"❌ 프로그램 실행 실패: {e}")


if __name__ == "__main__":
    main()

File: nodejs/energyNFTMinter.js
#!/usr/bin/env node
/**
 * 에너지 인증서 NFT 자동 민팅 서버 (Node.js + Express)
 * RE100, CBAM, 탄소거래용 NFT 생성 및 관리 API
 */

const express = require('express');
const { ethers } = require('ethers');
const axios = require('axios');
const FormData = require('form-data');
const fs = require('fs').promises;
const path = require('path');
const dotenv = require('dotenv');

// 환경 변수 로드
dotenv.config();

class EnergyNFTMinter {
    constructor() {
        this.provider = new ethers.JsonRpcProvider(process.env.RPC_URL);
        this.wallet = new ethers.Wallet(process.env.PRIVATE_KEY, this.provider);
        this.contractAddress = process.env.CONTRACT_ADDRESS;
        this.contract = null;
        this.initContract();
    }

    async initContract() {
        try {
            // ABI 파일 로드
            const abiPath = path.join(__dirname, '../artifacts/contracts/EnergyCertificateNFT.sol/EnergyCertificateNFT.json');
            const contractJson = JSON.parse(await fs.readFile(abiPath, 'utf8'));
            
            this.contract = new ethers.Contract(
                this.contractAddress,
                contractJson.abi,
                this.wallet
            );
            
            console.log('✅ 스마트컨트랙트 연결 완료');
        } catch (error) {
            console.error('❌ 컨트랙트 초기화 실패:', error);
        }
    }

    /**
     * 에너지 타입별 이미지 URL 반환
     */
    getEnergyTypeImage(energyType) {
        const imageMapping = {
            'solar': 'ipfs://QmSolarPanelIcon/solar.png',
            'wind': 'ipfs://QmWindTurbineIcon/wind.png',
            'hydro': 'ipfs://QmHydroIcon/hydro.png',
            'biomass': 'ipfs://QmBiomassIcon/biomass.png'
        };
        return imageMapping[energyType] || 'ipfs://QmDefaultEnergyIcon/default.png';
    }

    /**
     * 탄소 회피량 계산
     */
    calculateCarbonOffset(energyAmount, energyType) {
        const carbonFactors = {
            'solar': 0.466,
            'wind': 0.011,
            'hydro': 0.024,
            'biomass': 0.18
        };
        return energyAmount * (carbonFactors[energyType] || 0.4);
    }

    /**
     * NFT 메타데이터 생성
     */
    generateMetadata(energyData) {
        const carbonOffset = this.calculateCarbonOffset(
            energyData.energyAmount, 
            energyData.energyType
        );

        return {
            name: `RE100 전력 NFT #${energyData.tokenId || 'TBD'}`,
            description: `${energyData.energyAmount}MWh ${energyData.energyType} 전력 거래 인증서`,
            image: this.getEnergyTypeImage(energyData.energyType),
            external_url: `https://energy-nft.com/token/${energyData.tokenId || ''}`,
            attributes: [
                {
                    trait_type: "전력량 (MWh)",
                    value: energyData.energyAmount,
                    display_type: "number"
                },
                {
                    trait_type: "에너지 타입",
                    value: energyData.energyType
                },
                {
                    trait_type: "거래일시",
                    value: energyData.timestamp || new Date().toISOString()
                },
                {
                    trait_type: "공급자",
                    value: energyData.supplier
                },
                {
                    trait_type: "수요자",
                    value: energyData.buyer
                },
                {
                    trait_type: "발전소 위치",
                    value: energyData.location
                },
                {
                    trait_type: "탄소 회피량 (tCO₂)",
                    value: Math.round(carbonOffset * 1000) / 1000,
                    display_type: "number"
                },
                {
                    trait_type: "RE100 적합성",
                    value: "인증됨"
                },
                {
                    trait_type: "CBAM 대응",
                    value: "적합"
                }
            ],
            properties: {
                category: "Energy Certificate",
                subcategory: energyData.energyType.charAt(0).toUpperCase() + energyData.energyType.slice(1),
                is_renewable: true,
                certification_standard: "RE100",
                carbon_offset_tco2: carbonOffset,
                energy_amount_mwh: energyData.energyAmount
            }
        };
    }

    /**
     * IPFS에 메타데이터 업로드 (Pinata 사용)
     */
    async uploadToIPFS(metadata) {
        try {
            const formData = new FormData();
            const jsonBuffer = Buffer.from(JSON.stringify(metadata, null, 2));
            
            formData.append('file', jsonBuffer, {
                filename: `metadata_${Date.now()}.json`,
                contentType: 'application/json'
            });

            const response = await axios.post(
                'https://api.pinata.cloud/pinning/pinFileToIPFS',
                formData,
                {
                    headers: {
                        ...formData.getHeaders(),
                        'pinata_api_key': process.env.PINATA_API_KEY,
                        'pinata_secret_api_key': process.env.PINATA_SECRET_KEY
                    }
                }
            );

            const ipfsHash = response.data.IpfsHash;
            const ipfsUrl = `ipfs://${ipfsHash}`;
            
            console.log(`✅ IPFS 업로드 완료: ${ipfsUrl}`);
            return ipfsUrl;

        } catch (error) {
            console.error('❌ IPFS 업로드 실패:', error.message);
            throw error;
        }
    }

    /**
     * 에너지 NFT 민팅
     */
    async mintEnergyNFT(energyData) {
        try {
            console.log('🚀 NFT 민팅 시작...');

            // 1. 메타데이터 생성
            const metadata = this.generateMetadata(energyData);
            console.log('📄 메타데이터 생성 완료');

            // 2. IPFS 업로드
            const tokenURI = await this.uploadToIPFS(metadata);

            // 3. 탄소 회피량 계산 (정수로 변환)
            const carbonOffset = Math.floor(
                this.calculateCarbonOffset(energyData.energyAmount, energyData.energyType) * 1000
            );

            // 4. 에너지량을 kWh로 변환
            const energyAmountKWh = Math.floor(energyData.energyAmount * 1000);

            // 5. 스마트컨트랙트 함수 호출
            const tx = await this.contract.mintEnergyNFT(
                energyData.buyer,           // recipient
                tokenURI,                   // tokenURI
                energyAmountKWh,           // energyAmount (kWh)
                energyData.supplier,        // supplier
                energyData.buyer,          // buyer
                energyData.energyType,     // energyType
                carbonOffset,              // carbonOffset
                energyData.location,       // location
                {
                    gasLimit: 500000,
                    gasPrice: ethers.parseUnits('10', 'gwei')
                }
            );

            console.log(`✅ 민팅 트랜잭션 전송: ${tx.hash}`);

            // 6. 트랜잭션 확인 대기
            const receipt = await tx.wait();
            
            if (receipt.status === 1) {
                console.log(`🎉 NFT 민팅 성공! Gas 사용량: ${receipt.gasUsed}`);
                
                // 이벤트에서 토큰 ID 추출
                const event = receipt.logs.find(log => {
                    try {
                        return this.contract.interface.parseLog(log).name === 'EnergyNFTMinted';
                    } catch {
                        return false;
                    }
                });

                let tokenId = null;
                if (event) {
                    const parsedEvent = this.contract.interface.parseLog(event);
                    tokenId = parsedEvent.args.tokenId.toString();
                }

                return {
                    success: true,
                    txHash: tx.hash,
                    tokenId: tokenId,
                    gasUsed: receipt.gasUsed.toString(),
                    tokenURI: tokenURI,
                    metadata: metadata
                };
            } else {
                throw new Error('트랜잭션 실패');
            }

        } catch (error) {
            console.error('❌ NFT 민팅 실패:', error);
            return {
                success: false,
                error: error.message
            };
        }
    }

    /**
     * 배치 민팅
     */
    async batchMint(energyDataList) {
        const results = [];
        
        for (let i = 0; i < energyDataList.length; i++) {
            console.log(`📦 배치 민팅 진행 중... (${i + 1}/${energyDataList.length})`);
            
            const result = await this.mintEnergyNFT(energyDataList[i]);
            results.push({
                index: i,
                data: energyDataList[i],
                result: result
            });

            // API 레이트 리미트 방지
            if (i < energyDataList.length - 1) {
                await new Promise(resolve => setTimeout(resolve, 2000));
            }
        }

        return results;
    }

    /**
     * 토큰 정보 조회
     */
    async getTokenInfo(tokenId) {
        try {
            const energyData = await this.contract.getEnergyData(tokenId);
            const tokenURI = await this.contract.tokenURI(tokenId);
            const owner = await this.contract.ownerOf(tokenId);

            return {
                tokenId: tokenId,
                owner: owner,
                tokenURI: tokenURI,
                energyData: {
                    energyAmount: energyData.energyAmount.toString(),
                    timestamp: new Date(Number(energyData.timestamp) * 1000).toISOString(),
                    supplier: energyData.supplier,
                    buyer: energyData.buyer,
                    energyType: energyData.energyType,
                    carbonOffset: energyData.carbonOffset.toString(),
                    location: energyData.location,
                    isRetired: energyData.isRetired
                }
            };
        } catch (error) {
            throw new Error(`토큰 정보 조회 실패: ${error.message}`);
        }
    }

    /**
     * 총 탄소 회피량 조회
     */
    async getTotalCarbonOffset(ownerAddress) {
        try {
            const totalOffset = await this.contract.getTotalCarbonOffset(ownerAddress);
            return (Number(totalOffset) / 1000).toString(); // tCO2 단위로 변환
        } catch (error) {
            throw new Error(`탄소 회피량 조회 실패: ${error.message}`);
        }
    }
}

// Express 서버 설정
const app = express();
const PORT = process.env.PORT || 3000;

app.use(express.json());
app.use(express.static('public'));

// CORS 미들웨어
app.use((req, res, next) => {
    res.header('Access-Control-Allow-Origin', '*');
    res.header('Access-Control-Allow-Headers', 'Origin, X-Requested-With, Content-Type, Accept');
    res.header('Access-Control-Allow-Methods', 'GET, POST, PUT, DELETE');
    next();
});

// NFT 민터 인스턴스
const minter = new EnergyNFTMinter();

// API 엔드포인트
app.get('/', (req, res) => {
    res.json({
        message: '🌱 에너지 인증서 NFT 민팅 API',
        version: '1.0.0',
        endpoints: {
            mint: 'POST /api/mint',
            batchMint: 'POST /api/batch-mint',
            tokenInfo: 'GET /api/token/:tokenId',
            carbonOffset: 'GET /api/carbon-offset/:address'
        }
    });
});

// 단일 NFT 민팅
app.post('/api/mint', async (req, res) => {
    try {
        const energyData = req.body;
        
        // 필수 필드 검증
        const requiredFields = ['energyAmount', 'energyType', 'supplier', 'buyer', 'location'];
        for (const field of requiredFields) {
            if (!energyData[field]) {
                return res.status(400).json({
                    success: false,
                    error: `필수 필드 누락: ${field}`
                });
            }
        }

        const result = await minter.mintEnergyNFT(energyData);
        res.json(result);

    } catch (error) {
        res.status(500).json({
            success: false,
            error: error.message
        });
    }
});

// 배치 민팅
app.post('/api/batch-mint', async (req, res) => {
    try {
        const { energyDataList } = req.body;
        
        if (!Array.isArray(energyDataList) || energyDataList.length === 0) {
            return res.status(400).json({
                success: false,
                error: 'energyDataList는 비어있지 않은 배열이어야 합니다.'
            });
        }

        const results = await minter.batchMint(energyDataList);
        res.json({
            success: true,
            results: results
        });

    } catch (error) {
        res.status(500).json({
            success: false,
            error: error.message
        });
    }
});

// 토큰 정보 조회
app.get('/api/token/:tokenId', async (req, res) => {
    try {
        const tokenId = req.params.tokenId;
        const tokenInfo = await minter.getTokenInfo(tokenId);
        res.json({
            success: true,
            data: tokenInfo
        });

    } catch (error) {
        res.status(500).json({
            success: false,
            error: error.message
        });
    }
});

// 탄소 회피량 조회
app.get('/api/carbon-offset/:address', async (req, res) => {
    try {
        const address = req.params.address;
        const totalOffset = await minter.getTotalCarbonOffset(address);
        res.json({
            success: true,
            address: address,
            totalCarbonOffset: totalOffset,
            unit: 'tCO₂'
        });

    } catch (error) {
        res.status(500).json({
            success: false,
            error: error.message
        });
    }
});

// 서버 시작
app.listen(PORT, () => {
    console.log(`🚀 에너지 NFT 민팅 서버가 http://localhost:${PORT}에서 실행 중입니다.`);
    console.log(`📡 네트워크: ${process.env.NETWORK || 'sepolia'}`);
    console.log(`📍 컨트랙트: ${process.env.CONTRACT_ADDRESS}`);
});

module.exports = { EnergyNFTMinter, app };